cpu/native/syscalls.c

```diff
@@ -249,13 +249,18 @@ ssize_t _native_writev(int fd, const struct iovec *iov, int iovcnt)
     ssize_t r = 0;
 
     if (fd == STDOUT_FILENO || fd == STDERR_FILENO) {
-        for (int i = 0; i < iovcnt; ++i) {
-            ssize_t res = stdio_write(iov->iov_base, iov->iov_len);
-            iov++;
-            if (res < 0) {
-                return res;
+        while (iovcnt--) {
+            size_t offset = 0;
+            while (iov->iov_len > offset) {
+                ssize_t res = stdio_write((void *)((uintptr_t)iov->iov_base + offset),
+                                          iov->iov_len - offset);
+                if (res < 0) {
+                    return res;
```

Comment:
Is that how `writev` is supposed to behave oO?

Reply:
Manpage says:
```
On success, readv(), preadv(), and preadv2() return the number of bytes read; writev(), pwritev(), and pwritev2() return the number of bytes written.

       Note that it is not an error for a successful call to transfer fewer bytes than requested (see read(2) and write(2))
```

Reply:
[Spec](https://pubs.opengroup.org/onlinepubs/9699919799/functions/writev.html) is a bit different though.

Reply:
I am not sure about that since the the spec says on success which it does not define but the manpage does so. On the other hand the specs define the unsuccess (error) as the fd is not changed -> nothing is written?

Reply:
the spec says:
```
Upon successful completion, writev() shall return the number of bytes actually written.
```
but it does not define success,
but it defines the unsuccess (error) case 
```
 Otherwise, it shall return a value of -1, the file-pointer shall remain unchanged, and errno shall be set to indicate an error.
```
this says the the `file-pointer shall remain unchanged` on error i am not sure if there is a file-point file/char-device content equality but if there is this means the nothing has to be written if -1 if returned.

the manpage on the other hand defines the success as a partial write
```
Note that it is not an error for a successful call to transfer fewer bytes than requested
```

Reply:
the write manpage  goes into detail howto get the error state:
```
In  the  event  of  a  partial write, the caller can make another write() call to transfer the remaining bytes. \
The subsequent call will either transfer further bytes or may result in an error (e.g., if the disk is now full)
```